File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/cell/ThreadStatusCell.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.ui.cell;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Typeface;
import android.os.Handler;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.style.StyleSpan;
import android.util.AttributeSet;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.github.k1rakishou.chan.R;
import com.github.k1rakishou.chan.core.manager.BoardManager;
import com.github.k1rakishou.chan.core.model.ChanThread;
import com.github.k1rakishou.chan.core.model.Post;
import com.github.k1rakishou.chan.core.site.sites.chan4.Chan4PagesRequest;
import com.github.k1rakishou.chan.ui.theme.ThemeEngine;
import com.github.k1rakishou.chan.utils.AndroidUtils;
import com.github.k1rakishou.model.data.board.ChanBoard;
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor;

import javax.inject.Inject;

import static com.github.k1rakishou.chan.utils.AndroidUtils.getString;

public class ThreadStatusCell extends LinearLayout implements View.OnClickListener {
    private static final int UPDATE_INTERVAL = 1000;
    private static final int MESSAGE_INVALIDATE = 1;

    @Inject
    ThemeEngine themeEngine;
    @Inject
    BoardManager boardManager;

    private Callback callback;
    private boolean running = false;
    private TextView text;
    private String error;

    private Handler handler = new Handler(msg -> {
        if (msg.what == MESSAGE_INVALIDATE) {
            if (running && update()) {
                schedule();
            }

            return true;
        } else {
            return false;
        }
    });

    public ThreadStatusCell(Context context, AttributeSet attrs) {
        super(context, attrs);

        AndroidUtils.extractStartActivityComponent(context)
                .inject(this);

        setBackgroundResource(R.drawable.item_background);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        text = findViewById(R.id.text);
        text.setTypeface(themeEngine.getChanTheme().getMainFont());
        text.setTextColor(themeEngine.getChanTheme().getTextColorSecondary());

        setOnClickListener(this);
    }

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    public void setError(String error) {
        this.error = error;
        if (error == null) {
            schedule();
        }
    }

    @SuppressLint("SetTextI18n")
    public boolean update() {
        ChanThread chanThread = callback.getChanThread();
        if (chanThread == null) {
            // Recyclerview not clearing immediately or view didn't receive
            // onDetachedFromWindow.
            return false;
        }

        if (chanThread.getChanDescriptor().isCatalogDescriptor()) {
            if (isClickable()) {
                setClickable(false);
            }

            if (isFocusable()) {
                setFocusable(false);
            }

            return false;
        }

        if (error != null) {
            text.setText(error + "\n" + getString(R.string.thread_refresh_bar_inactive));
            return false;
        }

        boolean update = false;
        SpannableStringBuilder builder = new SpannableStringBuilder();

        if (chanThread.canUpdateThread()) {
            update = true;
        }

        boolean isNotDownloadedLocalThread = appendThreadStatusPart(chanThread, builder);
        if (isNotDownloadedLocalThread) {
            // To split Local Thread and (Loading Time | Loading) rows
            builder.append('\n');
        }

        appendThreadRefreshPart(chanThread, builder);

        // to split up the cell into the top and bottom rows
        builder.append('\n');

        Post op = chanThread.getOp();
        BoardDescriptor boardDescriptor = op.boardDescriptor;

        if (boardDescriptor != null) {
            ChanBoard board = boardManager.byBoardDescriptor(boardDescriptor);
            if (board != null) {
                appendThreadStatisticsPart(chanThread, builder, op, board);
            }
        }

        text.setText(builder);
        return update;
    }

    private void appendThreadStatisticsPart(ChanThread chanThread, SpannableStringBuilder builder, Post op, ChanBoard board) {
        boolean hasReplies = op.getTotalRepliesCount() >= 0 || chanThread.getPostsCount() - 1 > 0;
        boolean hasImages = op.getThreadImagesCount() >= 0 || chanThread.getImagesCount() > 0;

        if (hasReplies && hasImages) {
            boolean hasBumpLimit = board.getBumpLimit() > 0;
            boolean hasImageLimit = board.getImageLimit() > 0;

            SpannableString replies = new SpannableString(
                    (op.getTotalRepliesCount() >= 0 ? op.getTotalRepliesCount() : chanThread.getPostsCount() - 1) + "R");
            if (hasBumpLimit && op.getTotalRepliesCount() >= board.getBumpLimit()) {
                replies.setSpan(new StyleSpan(Typeface.ITALIC), 0, replies.length(), 0);
            }

            SpannableString images = new SpannableString(
                    (op.getThreadImagesCount() >= 0 ? op.getThreadImagesCount() : chanThread.getImagesCount()) + "I");
            if (hasImageLimit && op.getThreadImagesCount() >= board.getImageLimit()) {
                images.setSpan(new StyleSpan(Typeface.ITALIC), 0, images.length(), 0);
            }

            builder.append(replies).append(" / ").append(images);

            if (op.getUniqueIps() >= 0) {
                String ips = op.getUniqueIps() + "P";
                builder.append(" / ").append(ips);
            }
        }

        Chan4PagesRequest.BoardPage boardPage = callback.getPage(op);
        if (boardPage != null) {
            SpannableString page = new SpannableString(String.valueOf(boardPage.getCurrentPage()));
            if (boardPage.getCurrentPage() >= board.getPages()) {
                page.setSpan(new StyleSpan(Typeface.ITALIC), 0, page.length(), 0);
            }

            builder.append(" / ").append(getString(R.string.thread_page_no)).append(' ').append(page);
        }
    }

    private void appendThreadRefreshPart(ChanThread chanThread, SpannableStringBuilder builder) {
        if (!chanThread.canUpdateThread()) {
            return;
        }

        long time = callback.getTimeUntilLoadMore() / 1000L;
        if (!callback.isWatching()) {
            builder.append(getString(R.string.thread_refresh_bar_inactive));
        } else if (time <= 0) {
            builder.append(getString(R.string.loading));
        } else {
            builder.append(getString(R.string.thread_refresh_countdown, time));
        }
    }

    private boolean appendThreadStatusPart(ChanThread chanThread, SpannableStringBuilder builder) {
        if (chanThread.isArchived()) {
            builder.append(getString(R.string.thread_archived));
        } else if (chanThread.isClosed()) {
            builder.append(getString(R.string.thread_closed));
        } else if (chanThread.isDeleted()) {
            builder.append(getString(R.string.thread_deleted));
        }

        return false;
    }

    private void schedule() {
        running = true;

        if (!handler.hasMessages(MESSAGE_INVALIDATE)) {
            handler.sendMessageDelayed(handler.obtainMessage(MESSAGE_INVALIDATE), UPDATE_INTERVAL);
        }
    }

    private void unschedule() {
        running = false;
        handler.removeMessages(MESSAGE_INVALIDATE);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        schedule();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        unschedule();
    }

    @Override
    public void onWindowFocusChanged(boolean hasWindowFocus) {
        super.onWindowFocusChanged(hasWindowFocus);
        if (hasWindowFocus) {
            if (update()) {
                schedule();
            }
        } else {
            unschedule();
        }
    }

    @Override
    public void onClick(View v) {
        error = null;

        if (callback.getChanThread() != null) {
            callback.onListStatusClicked();
        }

        update();
    }

    public interface Callback {
        long getTimeUntilLoadMore();

        boolean isWatching();

        @Nullable
        ChanThread getChanThread();

        @Nullable
        Chan4PagesRequest.BoardPage getPage(Post op);

        void onListStatusClicked();
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/cell/ThreadStatusCell.kt
package com.github.k1rakishou.chan.ui.cell

import android.annotation.SuppressLint
import android.content.Context
import android.text.SpannableStringBuilder
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.github.k1rakishou.chan.R
import com.github.k1rakishou.chan.core.concurrency.RendezvousCoroutineExecutor
import com.github.k1rakishou.chan.core.concurrency.ThrottleFirstCoroutineExecutor
import com.github.k1rakishou.chan.core.manager.ArchivesManager
import com.github.k1rakishou.chan.core.manager.BoardManager
import com.github.k1rakishou.chan.core.manager.ChanThreadManager
import com.github.k1rakishou.chan.core.manager.ThreadDownloadManager
import com.github.k1rakishou.chan.ui.helper.AppResources
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils
import com.github.k1rakishou.common.isNotNullNorBlank
import com.github.k1rakishou.core_themes.ThemeEngine
import com.github.k1rakishou.model.data.board.pages.BoardPage
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor
import com.github.k1rakishou.model.data.descriptor.PostDescriptor
import com.github.k1rakishou.model.data.thread.ChanThread
import com.github.k1rakishou.model.data.thread.ThreadDownload
import com.github.k1rakishou.model.util.ChanPostUtils
import com.github.k1rakishou.v2.KurobaSettings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject

class ThreadStatusCell(
  context: Context,
  attrs: AttributeSet
) : LinearLayout(context, attrs), View.OnClickListener, ThemeEngine.ThemeChangesListener {

  @Inject
  lateinit var kurobaSettings: KurobaSettings
  @Inject
  lateinit var appResources: AppResources
  @Inject
  lateinit var themeEngine: ThemeEngine
  @Inject
  lateinit var boardManager: BoardManager
  @Inject
  lateinit var chanThreadManager: ChanThreadManager
  @Inject
  lateinit var archivesManager: ArchivesManager
  @Inject
  lateinit var threadDownloadManager: ThreadDownloadManager

  private lateinit var statusCellText: TextView

  private var callback: Callback? = null
  private var errorMessage: String? = null

  private val job = SupervisorJob()
  private val scope = CoroutineScope(job + Dispatchers.Main)
  private val rendezvousCoroutineExecutor = RendezvousCoroutineExecutor(scope)
  private val clickThrottler = ThrottleFirstCoroutineExecutor(scope)

  private var updateJob: Job? = null

  init {
    AppModuleAndroidUtils.extractActivityComponent(context)
      .inject(this)

    setBackgroundResource(R.drawable.item_background)
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()

    themeEngine.addListener(this)
    schedule()
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()

    themeEngine.removeListener(this)
    unschedule()
  }

  override fun onFinishInflate() {
    super.onFinishInflate()

    statusCellText = findViewById(R.id.text)
    statusCellText.typeface = themeEngine.chanTheme.mainFont

    setOnClickListener(this)
    onThemeChanged()
  }

  override fun onThemeChanged() {
    statusCellText.setTextColor(themeEngine.chanTheme.textColorSecondary)
  }

  override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
    super.onWindowFocusChanged(hasWindowFocus)

    if (hasWindowFocus) {
      schedule()
    } else {
      unschedule()
    }
  }

  private fun schedule() {
    unschedule()

    updateJob = scope.launch {
      update()
    }
  }

  private fun unschedule() {
    updateJob?.cancel()
    updateJob = null
  }

  override fun onClick(v: View) {
    clickThrottler.post(
      timeout = 1_000L,
      func = {
        errorMessage = null

        if (callback?.currentChanDescriptor != null) {
          callback?.onListStatusClicked()
        }

        update()
      }
    )
  }

  fun setCallback(callback: Callback?) {
    this.callback = callback
  }

  fun setError(error: String?) {
    this.errorMessage = error

    if (error == null) {
      schedule()
    }
  }

  fun update() {
    rendezvousCoroutineExecutor.post {
      updateInternal()
    }
  }

  @SuppressLint("SetTextI18n")
  suspend fun updateInternal() {
    val chanDescriptor = callback?.currentChanDescriptor
      ?: return

    if (chanDescriptor.isCatalogDescriptor()) {
      unschedule()
      updateCatalogStatusCell()
    } else {
      updateThreadStatusCell(chanDescriptor, callback)
    }
  }

  private fun updateCatalogStatusCell() {
    if (errorMessage != null) {
      val actualMessage = errorMessage.takeIf { it.isNotNullNorBlank() }
        ?: appResources.string(R.string.thread_status_cell_unknown_error)

      statusCellText.text = buildString {
        appendLine(appResources.string(R.string.catalog_refresh_error_text_title))
        appendLine("\"${actualMessage}\"")
        appendLine(appResources.string(R.string.catalog_click_to_refresh))
      }

      return
    }

    statusCellText.text = buildString {
      appendLine(appResources.string(R.string.catalog_refresh_title))
      appendLine(appResources.string(R.string.catalog_click_to_refresh))
    }
  }

  private suspend fun updateThreadStatusCell(
    chanDescriptor: ChanDescriptor,
    callback: Callback?
  ) {
    if (errorMessage != null) {
      val actualMessage = errorMessage.takeIf { it.isNotNullNorBlank() }
        ?: appResources.string(R.string.thread_status_cell_unknown_error)

      statusCellText.text = buildString {
        appendLine(appResources.string(R.string.thread_refresh_error_text_title))
        appendLine("\"${actualMessage}\"")
        appendLine(appResources.string(R.string.thread_refresh_bar_inactive))
      }

      return
    }

    chanDescriptor as ChanDescriptor.ThreadDescriptor

    if (chanThreadManager.isThreadLockCurrentlyLocked(chanDescriptor)) {
      // Since we update ThreadStatusCell every second there might be times when the ChanThread object
      // is being updated with new posts and there are a lot of posts so it may hold the lock for
      // quite some time. So to avoid freezing the whole app because of that we need to first check
      // whether the ChanThread lock is locked and skip this update if it's locked.
      return
    }

    val chanThread = chanThreadManager.getChanThread(chanDescriptor)
      ?: return

    val canUpdate = kurobaSettings.application.autoRefreshThread.read() && chanThread.canUpdateThread()
    val builder = SpannableStringBuilder()
      .apply { appendLine() }

    if (appendThreadStatusPart(chanThread, builder)) {
      builder.appendLine()
    }

    if (appendThreadRefreshPart(chanThread, builder)) {
      builder.appendLine()
    }

    val op = chanThread.getOriginalPost()
    if (op != null) {
      val board = boardManager.byBoardDescriptor(op.postDescriptor.boardDescriptor())
      val threadStatistics = ChanPostUtils.getThreadStatistics(
        chanThread = chanThread,
        op = op,
        board = board,
        boardPage = callback?.getPage(op.postDescriptor)
      )

      builder
        .append(threadStatistics)
        .appendLine()
    }

    if (archivesManager.isSiteArchive(chanDescriptor.siteDescriptor())) {
      builder
        .append(appResources.string(R.string.controller_bookmarks_bookmark_of_archived_thread))
    }

    appendThreadDownloaderStats(chanDescriptor, builder)

    builder.appendLine()

    statusCellText.text = builder

    if (canUpdate) {
      delay(UPDATE_INTERVAL_MS)
      schedule()
    }
  }

  private suspend fun appendThreadDownloaderStats(
    threadDescriptor: ChanDescriptor.ThreadDescriptor,
    builder: SpannableStringBuilder
  ) {
    when (threadDownloadManager.getStatus(threadDescriptor)) {
      ThreadDownload.Status.Running -> {
        builder
          .append(appResources.string(R.string.thread_status_downloading_running))
      }
      ThreadDownload.Status.Stopped -> {
        builder
          .append(appResources.string(R.string.thread_status_downloading_stopped))
      }
      ThreadDownload.Status.Completed -> {
        builder
          .append(appResources.string(R.string.thread_status_downloaded))
      }
      else -> {
        // no-op
      }
    }
  }

  private suspend fun appendThreadRefreshPart(
    chanThread: ChanThread,
    builder: SpannableStringBuilder
  ): Boolean {
    if (!chanThread.canUpdateThread()) {
      return false
    }

    val timeSeconds = callback?.timeUntilLoadMoreMs()?.div(1000L)
      ?: return false

    when {
      callback?.isWatching() == false -> {
        builder.append(appResources.string(R.string.thread_refresh_bar_inactive))
      }
      timeSeconds <= 0 -> {
        builder.append(appResources.string(R.string.loading))
      }
      else -> {
        builder.append(appResources.string(R.string.thread_refresh_countdown, timeSeconds))
      }
    }

    return true
  }

  private fun appendThreadStatusPart(
    chanThread: ChanThread,
    builder: SpannableStringBuilder
  ): Boolean {
    when {
      chanThread.isArchived() -> {
        builder.append(appResources.string(R.string.thread_archived))
        return true
      }
      chanThread.isClosed() -> {
        builder.append(appResources.string(R.string.thread_closed))
        return true
      }
      chanThread.isDeleted() -> {
        builder.append(appResources.string(R.string.thread_deleted))
        return true
      }
    }

    return false
  }

  interface Callback {
    val currentChanDescriptor: ChanDescriptor?

    suspend fun timeUntilLoadMoreMs(): Long
    fun isWatching(): Boolean
    suspend fun getPage(originalPostDescriptor: PostDescriptor): BoardPage?
    fun onListStatusClicked()
  }

  companion object {
    private const val UPDATE_INTERVAL_MS = 1000L
  }

}